src/test/scala/com/cloudant/clouseau/IndexServiceSpec.scala

```diff
@@ -25,10 +25,192 @@ import scalang.Pid
 import scala.Some
 import java.io.File
 import scala.collection.JavaConverters._
+import org.apache.lucene.queryparser.classic.ParseException
 
+/*
+mvn test -Dtest="com.cloudant.clouseau.IndexServiceSpec"
+*/
 class IndexServiceSpec extends SpecificationWithJUnit {
   sequential
 
+  "a groups parser" should {
+    def groups(value: Any): Map[Symbol, Any] = Map('groups -> value)
+
+    "return None if option map doesn't contain 'groups'" in {
+      IndexService.getGroups(Map().asInstanceOf[Map[Symbol, Any]]) must beEqualTo(None)
+    }
+
+    "return None if 'groups option contain 'nil" in {
+      IndexService.getGroups(groups('nil)) must beEqualTo(None)
+    }
+
+    "return correct result for empty list groups" in {
+      IndexService.getGroups(groups(List())) must beEqualTo(Some(List()))
+    }
+
+    "return correct result for groups in the expected format" in {
+      var input = List(
+        ("first", List(List())), ("second", List(Seq())), ("third", List("String")),
+        ('null, List(List())), ('null, List(Seq())), ('null, List("String"))
+      )
+      IndexService.getGroups(groups(input)) must beEqualTo(Some(List(
+        (Some("first"), List(List())), (Some("second"), List(Seq())), (Some("third"), List("String")),
+        (None, List(List())), (None, List(Seq())), (None, List("String"))
+      )))
+    }
+
+    "return error when groups is not a list" in {
+      IndexService.getGroups(groups("not a List")) must throwA[ParseException].like {
+        case e => e.getMessage must contain("invalid groups query")
+      }
+    }
+
+  }
+
+  "an include_fields parser" should {
+    def includeFields(value: Any): Map[Symbol, Any] = Map('include_fields -> value)
+
+    "return None if option map doesn't contain 'include_fields'" in {
+      IndexService.getIncludeFields(Map().asInstanceOf[Map[Symbol, Any]]) must beEqualTo(None)
+    }
+
+    "return None if 'include_fields option contain 'nil" in {
+      IndexService.getIncludeFields(includeFields('nil)) must beEqualTo(None)
+    }
+
+    "return correct result for empty list include_fields" in {
+      IndexService.getIncludeFields(includeFields(List())) must beEqualTo(Some(Set("_id")))
+    }
+
+    "return correct result for List[String]" in {
+      var planets = List("Mercury", "Venus", "Earth", "Mars", "Jupiter", "Saturn", "Uranus", "Neptune")
+      IndexService.getIncludeFields(includeFields(planets)) must beEqualTo(Some(
+        Set("Mercury", "Venus", "Earth", "Mars", "Jupiter", "Saturn", "Uranus", "Neptune", "_id")
```

Comment:
Can the order of elements be guaranteed in case of a `Set()`?

Reply:
Set looses information about the order. Therefore it doesn't matter in which order the planets are specified when Set is constructed using literal syntax. I could say the following and it would still work.

```scala
Set("_id", "Neptune", "Uranus", "Saturn", "Jupiter", "Mars", "Earth", "Venus", "Mercury")
```

The reason why Set is chosen is to be compatible with existent implementation see [here](https://github.com/cloudant-labs/clouseau/blob/master/src/main/scala/com/cloudant/clouseau/IndexService.scala#L290)

```scala
    val includeFields: Set[String] =
      request.options.getOrElse('include_fields, 'nil) match {
        case 'nil =>
          null
        case value: List[String] =>
          Set[String]() ++ ("_id" :: value).toSet
        case other =>
          throw new ParseException(other + " is not a valid include_fields query")
      }
```

Reply:
Right, makes sense.